### ISSUE TITLE
Better OSGi support for saaj-api

### ISSUE THREAD

**Issue by kriegfrj** [eclipse-ee4j/saaj-api#43]
JAX-WS relies heavily on `java.util.ServiceLoader` for pluggable implementations of the various JAX-WS APIs. The ServiceLoader does not work properly in an OSGi environment, and hence getting a SOAP client up-and-running using Jakarta saaj-api requires some hackery with the thread context classloader.

In order for soap-api to work properly in finding saaj implementations in an OSGi environment without TCCL hacks, it needs to be properly configured. There are a couple of possibilities:

## Use `org.glassfish.hk2.osgiresourcelocator`

This is a bundle that examines all bundles as they are loaded, looking to see if they have any files in their `META-INF/services` directory, and registering those that do. Clients can then call the utility methods in `org.glassfish.hk2.osgiresourcelocator` to find and instantiate them.

All of the other JAX-WS API components in Jakarta support OSGi in their respective `FactoryFinder` implementations using the `org.glassfish.hk2.osgiresourcelocator` bundle:

ws-api:
https://github.com/eclipse-ee4j/jax-ws-api/blob/9e05dfabc3a244a68604b143302d7eff5e7b8a07/api/src/main/java/javax/xml/ws/spi/FactoryFinder.java#L72-L75

bind-api:
https://github.com/eclipse-ee4j/jaxb-api/blob/55180e9082f0172019341dd5f8c44eabe42555cc/jaxb-api/src/main/java/javax/xml/bind/ContextFinder.java#L309-L310

However, it seems that the equivalent support is missing from the soap-api implementation.

## Use the OSGi ServiceLoader mediator standard

More information about this here: https://osgi.org/specification/osgi.cmpn/7.0.0/service.loader.html

Apache Aries SPI-Fly is the reference implementation: https://aries.apache.org/modules/spi-fly.html

## Pro/Con comparison of the two approaches

Arguments in favour of GlassFish's osgiresourcelocator implementation:
* Using the GlassFish osgiresourcelocator method is consistent with other Jakarta EE implementations like `bind-api` and `jax-ws-api`.
* As a result of the first point, this is easier to implement in the short-term.

Arguments in favour of the ServiceLoader mediator implementation:
* It is an OSGi standard rather than a proprietary one.
* Consumer bundles' Java code doesn't need to be explicitly written to support it - SPI-Fly can re-write them automatically. (Compare this with the GlassFish implementation, where a client must explicitly call the GlassFish module to find implementations in an OSGi context.)
* The Require-Capability/Provide-Capability syntax of the ServiceLoader mediator makes the dependencies explicit and understandable by the OSGi resolver module. This means that missing implementation modules will be discovered at bundle resolve time rather than at runtime. For example, if soap-api/saaj-api declared the following header"

`Require-Capability: osgi.serviceloader; filter:="(osgi.serviceloader=javax.xml.soap.SAAJMetaFactory)";cardinality:=multiple`

...then the OSGi resolver won't resolve the bundle (and hence won't start it) unless there is another bundle installed (eg, saaj-impl) with the corresponding `Provide-Capability` header declaring that it contains an implementation of `javax.xml.soap.SAAJMetaFactory`.

* Integrates with OSGi's in-built services infrastructure - ServiceLoader implementations are automatically registered as OSGi services in the framework, and benefit from all of the service management functionality that an OSGi framework provides.

## Conclusion

For what little my recommendation may be worth, I recommend the following:

1. Make it a medium-to-long-term goal to migrate all of the Jakarta API bundles and their reference implementations to use the ServiceLoader mediator specification, for all of the long-term benefits that provides.
2. Possibly in the short-term, add the few lines needed to `javax.xml.soap.FactoryFinder` to support loading SAAJMetaFactory implementations in an OSGi context without TCCL hacks.

As a bit of background, I feel strongly about using the ServiceLoader mediator approach in the long term, based on my own recent (and painful) experience in trying to get a SOAP client up-and-running in an OSGi container. I started with a simple "Hello World"-type implementation that I had working in plain Java, and tried to migrate it to OSGi. I had to keep running it only to get `ClassNotFoundException` errors at run time. I'd have to find the bundles containing the missing classes, add them, and try again. This took several iterations (plus a patch to soap-api in my working copy to use osgiresourcelocator) before I finally got a working client. Declaring the dependencies explicitly would make this process much easier and less error-prone.

**Comment by lukasj** [comment]
added org.glassfish.hk2.osgiresourcelocator so it is consistent with related APIs.

Keeping this open as a reminder about other (probably better) option.

**Comment by kriegfrj** [comment]
Thanks @lukasj! What's the time frame for the next release?

**Comment by lukasj** [comment]
@kriegfrj I see I forgot to reply here... API 1.4.2 is out for some time already and new release of implementation (1.5.2) [went out](https://www.eclipse.org/lists/metro-dev/msg00002.html) about a week ago

**Comment by kriegfrj** [comment]
> @kriegfrj I see I forgot to reply here... API 1.4.2 is out for some time already and new release of implementation (1.5.2) [went out](https://www.eclipse.org/lists/metro-dev/msg00002.html) about a week ago

Thanks @lukasj, i did notice that 1.4.2 had gone out thanks! Does 1.5.2 have the Service Loader Mediator support? :)

**Comment by lukasj** [comment]
@kriegfrj no, I'm [quite busy](https://www.eclipse.org/lists/metro-dev/msg00003.html) with moving everything to jakarta namespace

**Comment by kriegfrj** [comment]
> @kriegfrj no, I'm [quite busy](https://www.eclipse.org/lists/metro-dev/msg00003.html) with moving everything to jakarta namespace

:+1: It's ok, as a short term workaround @rotty3000 has made a patch to SPI-Fly so that you can manually override the serviceloader mediator's metadata for bundles that are missing it. So we can get around it now. But it will still be good if the bundle includes it because then it can help with the OSGi resolver ensuring that the right bundles are present.

**Comment by col-panic** [comment]
I'm a bit confused by the abundance of packages in mvncentral - which one is the correct one representing this saaj-api implementation in version 1.4.2?

**Comment by lukasj** [comment]
see https://eclipse-ee4j.github.io/metro-saaj/

**Comment by col-panic** [comment]
Thanks for the link, but I simply can't get this to work in the OSGI Equinox environment. @kriegfrj did you manage to use the osgiresourcelocator approach? Do I understand you right, that there is a spi-fly implementation that allows me to combine all of this?

**Comment by kriegfrj** [comment]
@col-panic , I did manage to get it to work using Equinox and osgiresourcelocator with the fix in this issue. Make sure you have the hk2 bundle installed and configured to start before the jax-ws bundles.

**Comment by col-panic** [comment]
@kriegfrj thx for the help. Unfortunately all of this does not play well. Application is Eclipse Equinox based with JAXB parts, serving and clienting in JAX-WS and serving and clienting in JAX-RS - transition to Java 11 with all of this seems not feasible at the moment.

**Comment by kriegfrj** [comment]
Hi @col-panic , with Java 11 i remember i needed to put an --add-module on the command line for some of the xml support... though I've never done server-side, my code was only for a client.

**Comment by lukasj** [comment]
What about following http://docs.osgi.org/reference/portable-java-contracts.html#references and use `JavaSOAP` for Contract ID? Would that be good enough? Note that that reference is missing the `javax.xml.soap` package/JSR 67. I'm also not sure if (and if so, then when) this list will adopt `jakarta` names.

**Comment by ajaylamba-provar** [comment]
I am also facing issues while using saaj-api in OSGi and getting errors like `javax.xml.soap.SOAPException: Unable to create message factory for SOAP: Unable to create SAAJ meta-factoryProvider com.sun.xml.messaging.saaj.soap.SAAJMetaFactoryImpl not found`. On debugging the MessageFactory code, it seems the context classloader does not have any class and hence the above error. Could someone please guide me to the steps to use saaj-api with Java 11?
Note: If I use the plain jars and add them to classpath, it works but does not work with plugin based dependency.
